app/WebParsing/ReqParser.hs

```diff
@@ -247,46 +288,107 @@ justParser = do
+courseOrProgParser :: Parser Req
+courseOrProgParser = Parsec.between Parsec.spaces Parsec.spaces $ Parsec.choice $ map Parsec.try [
     parParser,
     cutoffParser,
     justParser,
+    programOrParser,
     rawTextParser
     ]
 
+programParser :: Parser Req
+programParser = do
+    Parsec.spaces
+    program <- Parsec.manyTill Parsec.anyChar $ Parsec.choice $ map (Parsec.try . Parsec.lookAhead) [
+        degreeType,
+        programSuffix,
+        progGroupSeparator,
+        progOrSeparator,
+        Parsec.oneOf ".;" >> return "",
+        Parsec.eof >> return ""
+        ]
+    return $ PROGRAM program
+
+-- Flattens nested ORs into a single OR
+-- eg. OR [OR ["CS major, "Math major"], RAW "permission from instructor"]
+-- Nested ORs occur because the way programs are related through ORs is
+-- different than that of courses. So they each have their orParser, which
+-- may be related throuhg another OR
+flattenOr :: [Req] -> [Req]
+flattenOr [] = []
+flattenOr (OR x:xs) = x ++ flattenOr xs
+flattenOr (x:xs) = x:flattenOr xs
+
 -- | Parser for reqs related through the "one of the following" condition
 -- | Courses that fall under the one of condition may be separated by orSeparators or andSeparators
 oneOfParser :: Parser Req
 oneOfParser = do
     Parsec.spaces
     _ <- Parsec.try oneOfSeparator
     Parsec.spaces
-    reqs <- Parsec.sepBy courseParser (Parsec.try orSeparator <|> Parsec.try andSeparator)
+    reqs <- Parsec.sepBy courseOrProgParser (Parsec.try orSeparator <|> Parsec.try andSeparator)
     case reqs of
         [] -> fail "Empty Req."
         [x] -> return x
-        (x:xs) -> return $ OR (x:xs)
+        (x:xs) -> return $ OR $ flattenOr (x:xs)
 
 -- | Parser for reqs related through an OR.
 orParser :: Parser Req
 orParser = do
-    reqs <- Parsec.sepBy courseParser orSeparator
+    reqs <- Parsec.sepBy courseOrProgParser orSeparator
     case reqs of
         [] -> fail "Empty Req."
         [x] -> return x
-        (x:xs) -> return $ OR (x:xs)
+        (x:xs) -> return $ OR $ flattenOr (x:xs)
 
 -- | Parser for for reqs related through an AND.
 andParser :: Parser Req
 andParser = do
-    reqs <- Parsec.sepBy (Parsec.try oneOfParser <|> orParser) andSeparator
+    reqs <- Parsec.sepBy (Parsec.try oneOfParser <|> orParser <|> programOrParser) andSeparator
     case reqs of
         [] -> fail "Empty Req."
         [x] -> return x
         (x:xs) -> return $ AND (x:xs)
 
+-- | Parser for programs grouped together
+-- | Parses program names and degree types, then concatenate every combination
+-- | eg. (CS or MAT major) implies (CS major) or (MAT major)
+-- | (CS major or specialist) implies (CS major) or (CS specialist)
+-- | (CS or MAT major or specialist) implies (CS maj) or (MAT maj) or (CS spec) or (MAT spec)
+programGroupParser :: Parser Req
+programGroupParser = do
+    progs <- Parsec.sepBy (Parsec.try programParser) progGroupSeparator
+    degrees <- Parsec.sepBy (Parsec.try degreeType) (Parsec.string "or")
```

Comment:
The `or` literal here is a bit strange; seems like this needs to be combined with some of the other parsers you defined above for "or" separators.

Reply:
The problem I had here was that the degree types were only ever separated by `or`s, and other separators in `orSeparator` and `programOrSeparator` would overmatch. For example, the `, or` in `orSeparator` would split "Enrolment in a History major, or permission of instructor" (because we're using `Parsec.sepBy`) and would try to read the "permission of instructor" as a degree type, which would then fail because it is not one of "major", "specialist", "program"... etc.

Should I define a `degreeOrParser` (similar to `programOrParser`) that matches only `or`s and reference it here to make it more clear?

Reply:
In general we shouldn't rely on different separators like this. The parser should also work on `Enrolment in a History major or permission of instructor`. (See above comment about overfitting.)

Seems like the problem is failing parsers consuming characters we don't want to be consumed. (i.e., the `or` being consumed even if the `degreeType` parser fails.) If this is the case, then it seems like we'll need to do something more sophisticated. You can take a look at [how `sepBy` is implemented](https://hackage.haskell.org/package/parsec-3.1.15.1/docs/src/Text.Parsec.Combinator.html#sepBy) to get some ideas about implementing something similar, perhaps.